File: 8-numberOfIslands.py
"""
Count the number of ilands in a 2 dimentional matrix

A iland is defined to a portio

example:
[
    [1,1,1,1,1,1,1,1,1],
    [1,1,1,1,1,1,1,1,1],
    [1,1,0,0,0,0,0,1,1],
    [1,1,0,0,0,0,0,1,1],
    [1,1,0,0,0,0,0,1,1],
    [1,1,0,0,0,0,0,0,0],
]

result should be 1
"""

from operator import index


island_map = [
    [0,1,1,1,1,1,1,1,1],
    [1,0,1,1,1,1,1,1,1],
    [1,1,0,0,0,0,0,1,1],
    [1,1,0,0,0,0,0,1,1],
    [1,1,0,0,1,1,0,1,1],
    [1,1,0,0,1,1,0,0,0]
]

# result shoud be 1

def countHowManyIslands(matrix_map):
    mutable_matrix_map = matrix_map
    count = 0

    def depthFirstSearch(i, j):
        
        # Index out of bounds
        if i < 0 or i > len(mutable_matrix_map) - 1:
            return
        if j < 0 or j > len(mutable_matrix_map[i]) - 1:
            return

        # Found the limit of the island, return
        if mutable_matrix_map[i][j] == 0:
            return

        # Found land, change to 0 to not count twice
        if mutable_matrix_map[i][j] == 1:
            mutable_matrix_map[i][j] = 0

        # Search all sides
        depthFirstSearch(i + 1, j)
        depthFirstSearch(i - 1, j)
        depthFirstSearch(i, j + 1)
        depthFirstSearch(i, j - 1)

        # Return 1 to count as 1 island
        return 1

    # BEFORE
    for i in range(0, len(mutable_matrix_map) - 1):
        for j in range(0, len(mutable_matrix_map[i]) - 1):
            print(mutable_matrix_map[i][j], end="")
        print("")

    # PROCESS
    for i in range(0, len(mutable_matrix_map) - 1):
        for j in range(0, len(mutable_matrix_map[i]) - 1):
            if mutable_matrix_map[i][j] == 1:
                count += depthFirstSearch(i, j)

    # AFTER
    for i in range(0, len(mutable_matrix_map) - 1):
        for j in range(0, len(mutable_matrix_map[i]) - 1):
            print(mutable_matrix_map[i][j], end="")
        print("")

    print("Number of islands = ", count)


        


countHowManyIslands(island_map)

File: 8-numberOfIslands.swift
// Count the number of islands in a 2 dimentional matrix

// An island is defined to a portion of land represented by 1
// Water is represented as 0

// example:
// [
//     [1,1,1,1,1,1,1,1,1],
//     [1,1,1,1,1,1,1,1,1],
//     [1,1,0,0,0,0,0,1,1],
//     [1,1,0,0,0,0,0,1,1],
//     [1,1,0,0,0,0,0,1,1],
//     [1,1,0,0,0,0,0,0,0],
// ]

// result should be 1

// THIS IS A SOLUTION FOR ITERATIVE APPROACH USING A LIST

var landMap = [
    [1,1,1,1,1,1,1,1,1],
    [1,1,1,1,1,1,1,1,1],
    [1,1,0,0,0,0,0,1,1],
    [1,1,0,1,1,0,0,1,1],
    [1,1,0,0,1,1,0,0,1],
    [1,1,0,0,0,0,0,0,0]
]

// SHOW MAP
for i in landMap.indices {
    let i = Int(i)
    for j in landMap[i].indices {
        let j = Int(j)
        print("\(landMap[i][j])", separator: " ", terminator: "")
    }
    print("")
}

struct Tuple: Hashable {
    let x: Int
    let y: Int
}

func countNumberOfIslands(in island_map: [[Int]]) -> Int {

    var stack = Set<Tuple>() {
        didSet {
            print("Stack = \(stack)")
        }
    }
    var map = island_map
    var counter = 0

    for i in map.indices {
        let i = Int(i)
        for j in map[i].indices {
            let j = Int(j)
            if map[i][j] == 1 {
                stack.insert(Tuple(x: i, y: j))
                // SHOW MAP
                print("")
                for i in map.indices {
                    let i = Int(i)
                    for j in map[i].indices {
                        let j = Int(j)
                        print("\(map[i][j])", separator: " ", terminator: "")
                    }
                    print("")
                }
                while !stack.isEmpty {
                    let tuple = stack.removeFirst()
                    let new_i = tuple.x
                    let new_j = tuple.y
                    map[new_i][new_j] = 0
                    //left
                    if new_i > 0 {
                        if map[new_i - 1][new_j] == 1 {
                            stack.insert(Tuple(x: new_i - 1, y: new_j))
                        }
                    }
                    //up
                    if new_j > 0 {
                        if map[new_i][new_j - 1] == 1 {
                            stack.insert(Tuple(x: new_i, y: new_j - 1))
                        }
                    }
                    //right
                    if new_i < map.count - 1 {
                        if map[new_i + 1][new_j] == 1 {
                            stack.insert(Tuple(x: new_i + 1, y: new_j))
                        }
                    }
                    //down
                    if new_j < map[new_i].count - 1 {
                        if map[new_i][new_j + 1] == 1 {
                            stack.insert(Tuple(x: new_i, y: new_j + 1))
                        }
                    }
                }
                counter += 1
                print("Counter = \(counter)")
                
            }
        }
    }
    return counter
}



let result = countNumberOfIslands(in: landMap)
print("Number os islands is = \(result)")